packages/core/src/validate/run.ts

```diff
@@ -261,24 +285,31 @@ function* getContractOpcodeErrors(
   decodeSrc: SrcDecoder,
   opcode: OpcodePattern,
   scope: Scope,
-  visitedNodeIds = new Map<number, Scope>(),
+  cache: Cache,
 ): Generator<ValidationErrorOpcode> {
-  if (wasVisited(contractDef.id, scope, visitedNodeIds)) {
-    // We return early here, but the errors from this node were emitted when the node was marked as visited.
-    return;
+  if (wasVisited(contractDef.id, scope, cache.visitedNodeIds)) {
+    yield* getCached(contractDef.id, scope, cache);
   } else {
-    visitedNodeIds.set(contractDef.id, scope);
+    cache.visitedNodeIds.set(contractDef.id, scope);
+    const result = [
+      ...getFunctionOpcodeErrors(contractDef, deref, decodeSrc, opcode, scope, cache),
+      ...getInheritedContractOpcodeErrors(contractDef, deref, decodeSrc, opcode, cache),
+    ];
+    yield* cacheAndYieldResult(contractDef.id, scope, cache, result);
```

Comment:
Now that we have separate caches for inherited and main, we may not need the `visitedNodeIds` set any more. I'll propose an alternative but let me know what you think is best.

We can just use the caches directly.

```typescript
    const cached = cache.[main or inherited].get(contractDef.id);
    if (cached !== undefined) {
      yield* cached;
    } else {
      const errors = [];
      cache.[main or inherited].set(contractDef.id, errors);
      errors.push(
        ...getFunctionOpcodeErrors(contractDef, deref, decodeSrc, opcode, scope, cache),
        ...getInheritedContractOpcodeErrors(contractDef, deref, decodeSrc, opcode, cache),
      );
      yield* errors;
    }
```

If there is a cache hit, we yield that. If there is a cache miss, we create an empty array and store it in the cache, then we get the errors and push them into the cached array. Any recursive cache hits will yield that array but before any errors are pushed into it.

I've implement this pattern a few times so I'm familiar with it but what do you think?

Reply:
This makes sense since the empty array is a simpler equivalent to the code before, where it had a visitedNodeId that was set but did not have anything in the cached errors.